Commit "Optimize calculation of element coordinates" broke PhantomJS compatibility

Relating to a discussion starting at https://github.com/pazone/ashot/pull/133#issuecomment-742951302, I want to provide more details. But first let me repeat the basic information from there:

> I just switched from 1.5.4 to 1.6.0-SNAPSHOT (not nice because I cannot release with snapshot dependencies in my POM).
>   * It solves an IE issue similar to #70 and #110 and my sample test also works in Chrome (also headless), Opera, Firefox, Edge (Chromium), just like before. So far, so good.
>   * But in 1.5.4 the same test used to run in PhantomJS, now with this version it fails.

Later in the discussion I offered:

> Does my hint that AShot 1.5.4 works with PJS, but the current snapshot does not, ring any bells? What has changed since then? Maybe you already have an idea. Otherwise, would it help to know the exact commit which broke PJS? Maybe I can `git bisect` it then.

The latter is what I did. I bisected from 1.5.4 to the current HEAD and found the exact commit which broke PhantomJS support. If you happen to not know what `git bisect` does, you may want to read [here](https://git-scm.com/docs/git-bisect). Anyway, here is my bisection protocol:

```text
$ git bisect bad
1c80c9bab3c87d3985601a7fd98c8e78a2d937fb is the first bad commit
commit 1c80c9bab3c87d3985601a7fd98c8e78a2d937fb
Author: Valery Yatsynovich <valfirst@yandex.ru>
Date:   Sat Dec 14 01:18:18 2019 +0300

    Optimize calculation of element coordinates

:040000 040000 4be584c1face43112a2037bd8d53e15630d31374 94f8da9076a6424a93fbc67625e1da1926ec0909 M      src

$ git bisect log
# bad: [885fa50dc459ff6dab730984553d4ed018e13a30] Merge pull request #199 from valfirst/bump-selenium
# good: [e6e5288ed3fe3304835f4cb95bccdbaba9a10cab] [maven-release-plugin] prepare release ashot-1.5.4
git bisect start 'HEAD' 'ashot-1.5.4'
# good: [c79058997ae66e19406fcfa4416e4dab5f741f8f] Merge pull request #168 from valfirst/remove-oss-parent-dependency
git bisect good c79058997ae66e19406fcfa4416e4dab5f741f8f
# bad: [130798e44f10429bed6e393ffbd3bdec1dee462b] Merge pull request #183 from valfirst/fix-ie-compatibility
git bisect bad 130798e44f10429bed6e393ffbd3bdec1dee462b
# bad: [208c883bb911ae56b0a41bcebecb80accfe9ebbf] Merge pull request #174 from valfirst/optimize-extraction-of-page-dimensions
git bisect bad 208c883bb911ae56b0a41bcebecb80accfe9ebbf
# bad: [d61b8b871f88d2e6ab4fe41c46686a3bec6c5867] Merge pull request #178 from valfirst/bump-mockito
git bisect bad d61b8b871f88d2e6ab4fe41c46686a3bec6c5867
# bad: [3fe877a93c6c829b8c8ad38c26016258ad6a9351] Merge pull request #179 from valfirst/optimize-calculation-of-element-coordinates
git bisect bad 3fe877a93c6c829b8c8ad38c26016258ad6a9351
# bad: [1c80c9bab3c87d3985601a7fd98c8e78a2d937fb] Optimize calculation of element coordinates
git bisect bad 1c80c9bab3c87d3985601a7fd98c8e78a2d937fb
# first bad commit: [1c80c9bab3c87d3985601a7fd98c8e78a2d937fb] Optimize calculation of element coordinates
```

During bisection I saw two different types of exception in my test. Mostly it was

```text
org.openqa.selenium.UnsupportedCommandException: Invalid Command Method - {"headers":{"Accept-Encoding":"gzip","Cache-Control":"no-cache","Connection":"Keep-Alive","Host":"localhost:10142","User-Agent":"selenium/3.14.0 (java windows)"},"httpVersion":"1.1","method":"GET","url":"/rect","urlParsed":{"anchor":"","query":"","file":"rect","directory":"/","path":"/rect","relative":"/rect","port":"","host":"","password":"","user":"","userInfo":"","authority":"","protocol":"","source":"/rect","queryKey":{},"chunks":["rect"]},"urlOriginal":"/session/70bd7c00-3c18-11eb-8458-cdcd98f57381/element/:wdc:1607736129298/rect"}
Command duration or timeout: 15 milliseconds
Build info: version: '3.14.0', revision: 'aacccce0', time: '2018-08-02T20:19:58.91Z'
System info: host: 'XANDER-ULTRABOO', ip: '192.168.137.1', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '1.8.0_211'
Driver info: org.openqa.selenium.phantomjs.PhantomJSDriver
Capabilities {acceptSslCerts: false, applicationCacheEnabled: false, browserConnectionEnabled: false, browserName: phantomjs, cssSelectorsEnabled: true, databaseEnabled: false, driverName: ghostdriver, driverVersion: 1.2.0, handlesAlerts: false, javascriptEnabled: true, locationContextEnabled: false, nativeEvents: true, platform: XP, platformName: XP, proxy: Proxy(direct), rotatable: false, takesScreenshot: true, version: 2.1.1, webStorageEnabled: false}
Session ID: 70bd7c00-3c18-11eb-8458-cdcd98f57381

	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:214)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:166)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:582)
	at org.openqa.selenium.remote.RemoteWebElement.execute(RemoteWebElement.java:276)
	at org.openqa.selenium.remote.RemoteWebElement.getRect(RemoteWebElement.java:345)
	at pazone.ashot.coordinates.WebDriverCoordsProvider.ofElement(WebDriverCoordsProvider.java:13)
	at pazone.ashot.coordinates.CoordsProvider.ofElements(CoordsProvider.java:24)
	at pazone.ashot.AShot.takeScreenshot(AShot.java:112)
	at de.scrum_master.testing.AShotReportingSpec.reportAShot(AShotReportingSpec.groovy:53)
	at de.scrum_master.testing.AShotReportingSpec.reportElements(AShotReportingSpec.groovy:42)
	at de.scrum_master.testing.AShotIT.Visit Scrum-Master.de download page(AShotIT.groovy:12)
```

This occurrect not during taking a full page screenshot (that one worked) but a single element screenshot with PhantomJS.

But at some point I also saw an even worse exception when trying to take a simple full page screenshot:

```text
java.lang.RuntimeException: org.openqa.selenium.WebDriverException: {"errorMessage":"Unexpected token ','","request":{"headers":{"Accept-Encoding":"gzip","Connection":"Keep-Alive","Content-Length":"519","Content-Type":"application/json; charset=utf-8","Host":"localhost:18071","User-Agent":"selenium/3.14.0 (java windows)"},"httpVersion":"1.1","method":"POST","post":"{\n  \"script\": \"var body = document.body;\\r\\nvar documentElement = document.documentElement;\\r\\nvar pageHeight = Math.max(body.scrollHeight, body.offsetHeight, documentElement.clientHeight,\\r\\n    documentElement.scrollHeight, documentElement.offsetHeight);\\r\\nvar viewportHeight = window.innerHeight || documentElement.clientHeight|| body.clientHeight;\\r\\nvar viewportWidth = window.innerWidth || documentElement.clientWidth || body.clientWidth;\\r\\nreturn {pageHeight, viewportHeight, viewportWidth}\",\n  \"args\": [\n  ]\n}","url":"/execute","urlParsed":{"anchor":"","query":"","file":"execute","directory":"/","path":"/execute","relative":"/execute","port":"","host":"","password":"","user":"","userInfo":"","authority":"","protocol":"","source":"/execute","queryKey":{},"chunks":["execute"]},"urlOriginal":"/session/8bd02670-3c1b-11eb-ab15-0fb31fb19d93/execute"}}
Command duration or timeout: 0 milliseconds

	at pazone.ashot.util.InnerScript.execute(InnerScript.java:28)
	at pazone.ashot.ViewportPastingDecorator.getPageDimensions(ViewportPastingDecorator.java:69)
	at pazone.ashot.ViewportPastingDecorator.getScreenshot(ViewportPastingDecorator.java:43)
	at pazone.ashot.AShot.takeScreenshot(AShot.java:113)
	at de.scrum_master.testing.AShotReportingSpec.reportAShot(AShotReportingSpec.groovy:53)
	at de.scrum_master.testing.AShotReportingSpec.reportFullPage(AShotReportingSpec.groovy:38)
	at de.scrum_master.testing.AShotIT.Visit Scrum-Master.de download page(AShotIT.groovy:11)
Caused by: org.openqa.selenium.WebDriverException: {"errorMessage":"Unexpected token ','","request":{"headers":{"Accept-Encoding":"gzip","Connection":"Keep-Alive","Content-Length":"519","Content-Type":"application/json; charset=utf-8","Host":"localhost:18071","User-Agent":"selenium/3.14.0 (java windows)"},"httpVersion":"1.1","method":"POST","post":"{\n  \"script\": \"var body = document.body;\\r\\nvar documentElement = document.documentElement;\\r\\nvar pageHeight = Math.max(body.scrollHeight, body.offsetHeight, documentElement.clientHeight,\\r\\n    documentElement.scrollHeight, documentElement.offsetHeight);\\r\\nvar viewportHeight = window.innerHeight || documentElement.clientHeight|| body.clientHeight;\\r\\nvar viewportWidth = window.innerWidth || documentElement.clientWidth || body.clientWidth;\\r\\nreturn {pageHeight, viewportHeight, viewportWidth}\",\n  \"args\": [\n  ]\n}","url":"/execute","urlParsed":{"anchor":"","query":"","file":"execute","directory":"/","path":"/execute","relative":"/execute","port":"","host":"","password":"","user":"","userInfo":"","authority":"","protocol":"","source":"/execute","queryKey":{},"chunks":["execute"]},"urlOriginal":"/session/8bd02670-3c1b-11eb-ab15-0fb31fb19d93/execute"}}
Command duration or timeout: 0 milliseconds
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:214)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:166)
	at org.openqa.selenium.remote.http.JsonHttpResponseCodec.reconstructValue(JsonHttpResponseCodec.java:40)
	at org.openqa.selenium.remote.http.AbstractHttpResponseCodec.decode(AbstractHttpResponseCodec.java:80)
	at org.openqa.selenium.remote.http.AbstractHttpResponseCodec.decode(AbstractHttpResponseCodec.java:44)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:158)
	at org.openqa.selenium.phantomjs.PhantomJSCommandExecutor.execute(PhantomJSCommandExecutor.java:82)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:548)
	at org.openqa.selenium.remote.RemoteWebDriver.executeScript(RemoteWebDriver.java:485)
	at pazone.ashot.util.InnerScript.execute(InnerScript.java:26)
	... 6 more
Caused by: org.openqa.selenium.remote.ScreenshotException: Screen shot has been taken
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:140)
	... 14 more
Caused by: org.openqa.selenium.WebDriverException: {"errorMessage":"Unexpected token ','","request":{"headers":{"Accept-Encoding":"gzip","Connection":"Keep-Alive","Content-Length":"519","Content-Type":"application/json; charset=utf-8","Host":"localhost:18071","User-Agent":"selenium/3.14.0 (java windows)"},"httpVersion":"1.1","method":"POST","post":"{\n  \"script\": \"var body = document.body;\\r\\nvar documentElement = document.documentElement;\\r\\nvar pageHeight = Math.max(body.scrollHeight, body.offsetHeight, documentElement.clientHeight,\\r\\n    documentElement.scrollHeight, documentElement.offsetHeight);\\r\\nvar viewportHeight = window.innerHeight || documentElement.clientHeight|| body.clientHeight;\\r\\nvar viewportWidth = window.innerWidth || documentElement.clientWidth || body.clientWidth;\\r\\nreturn {pageHeight, viewportHeight, viewportWidth}\",\n  \"args\": [\n  ]\n}","url":"/execute","urlParsed":{"anchor":"","query":"","file":"execute","directory":"/","path":"/execute","relative":"/execute","port":"","host":"","password":"","user":"","userInfo":"","authority":"","protocol":"","source":"/execute","queryKey":{},"chunks":["execute"]},"urlOriginal":"/session/8bd02670-3c1b-11eb-ab15-0fb31fb19d93/execute"}}
Build info: version: '3.14.0', revision: 'aacccce0', time: '2018-08-02T20:19:58.91Z'
System info: host: 'XANDER-ULTRABOO', ip: '192.168.137.1', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '1.8.0_211'
Driver info: driver.version: unknown
```

Both exceptions are reproducible for the corresponding commits, but the first one is the one currently seen in master.

I.e., the seemingly harmless commit https://github.com/pazone/ashot/commit/1c80c9bab3c87d3985601a7fd98c8e78a2d937fb broke PhantomJS compatibility. It "just" switches from `Point` and `Dimension` for WebDriver coordinate calculation to `Rectangle`, and PhantomJS does not seem to like it. Would you mind reverting that? I tried that on master and it works for me. If you do not want to lose the whole commit which also introduced a unit test, of course you would only revert `WebDriverCoordsProvider` and then adjust `WebDriverCoordsProviderTest` to match the reverted behaviour. Please let me know if you can simply do that as a maintainer or if this small change requires me to provide a PR including updated unit test.

--- a/src/main/java/pazone/ashot/coordinates/WebDriverCoordsProvider.java
+++ b/src/main/java/pazone/ashot/coordinates/WebDriverCoordsProvider.java
@@ -1,8 +1,10 @@
 package pazone.ashot.coordinates;
 
-import org.openqa.selenium.Rectangle;
+import org.openqa.selenium.Dimension;
+import org.openqa.selenium.Point;
 import org.openqa.selenium.WebDriver;
 import org.openqa.selenium.WebElement;
+
 
 /**
  * @author <a href="pazone@yandex-team.ru">Pavel Zorin</a>
@@ -10,11 +12,12 @@ import org.openqa.selenium.WebElement;
 public class WebDriverCoordsProvider extends CoordsProvider {
     @Override
     public Coords ofElement(WebDriver driver, WebElement element) {
-        Rectangle rect = element.getRect();
+        Point point = element.getLocation();
+        Dimension dimension = element.getSize();
         return new Coords(
-                rect.getX(),
-                rect.getY(),
-                rect.getWidth(),
-                rect.getHeight());
+                point.getX(),
+                point.getY(),
+                dimension.getWidth(),
+                dimension.getHeight());
     }
 }

--- a/src/main/java/pazone/ashot/coordinates/WebDriverCoordsProvider.java
+++ b/src/main/java/pazone/ashot/coordinates/WebDriverCoordsProvider.java
@@ -9,9 +9,11 @@ import org.openqa.selenium.WebElement;
 /**
  * @author <a href="pazone@yandex-team.ru">Pavel Zorin</a>
  */
+
 public class WebDriverCoordsProvider extends CoordsProvider {
     @Override
     public Coords ofElement(WebDriver driver, WebElement element) {
+        // Keep Point/Dimension as opposed to Rectangle because browsers like PhantomJS do not like the latter
         Point point = element.getLocation();
         Dimension dimension = element.getSize();
         return new Coords(

--- a/src/main/java/pazone/ashot/coordinates/WebDriverCoordsProvider.java
+++ b/src/main/java/pazone/ashot/coordinates/WebDriverCoordsProvider.java
@@ -5,11 +5,9 @@ import org.openqa.selenium.Point;
 import org.openqa.selenium.WebDriver;
 import org.openqa.selenium.WebElement;
 
-
 /**
  * @author <a href="pazone@yandex-team.ru">Pavel Zorin</a>
  */
-
 public class WebDriverCoordsProvider extends CoordsProvider {
     @Override
     public Coords ofElement(WebDriver driver, WebElement element) {

--- a/src/test/java/pazone/ashot/coordinates/WebDriverCoordsProviderTest.java
+++ b/src/test/java/pazone/ashot/coordinates/WebDriverCoordsProviderTest.java
@@ -1,14 +1,14 @@
 package pazone.ashot.coordinates;
 
-import org.junit.jupiter.api.Test;
-import org.openqa.selenium.Dimension;
-import org.openqa.selenium.Point;
-import org.openqa.selenium.WebElement;
-
 import static org.junit.jupiter.api.Assertions.assertAll;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
+
+import org.junit.jupiter.api.Test;
+import org.openqa.selenium.Dimension;
+import org.openqa.selenium.Point;
+import org.openqa.selenium.WebElement;
 
 class WebDriverCoordsProviderTest {
     private WebDriverCoordsProvider webDriverCoordsProvider = new WebDriverCoordsProvider();
@@ -21,7 +21,7 @@ class WebDriverCoordsProviderTest {
         int height = 3;
         int width = 4;
         when(element.getLocation()).thenReturn(new Point(x, y));
-        when(element.getSize()).thenReturn(new Dimension(4, 3));
+        when(element.getSize()).thenReturn(new Dimension(width, height));
         Coords coords = webDriverCoordsProvider.ofElement(null, element);
         assertAll(
                 () -> assertEquals(x, coords.x),

--- a/src/test/java/pazone/ashot/coordinates/WebDriverCoordsProviderTest.java
+++ b/src/test/java/pazone/ashot/coordinates/WebDriverCoordsProviderTest.java
@@ -1,13 +1,14 @@
 package pazone.ashot.coordinates;
 
+import org.junit.jupiter.api.Test;
+import org.openqa.selenium.Dimension;
+import org.openqa.selenium.Point;
+import org.openqa.selenium.WebElement;
+
 import static org.junit.jupiter.api.Assertions.assertAll;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
-
-import org.junit.jupiter.api.Test;
-import org.openqa.selenium.Rectangle;
-import org.openqa.selenium.WebElement;
 
 class WebDriverCoordsProviderTest {
     private WebDriverCoordsProvider webDriverCoordsProvider = new WebDriverCoordsProvider();
@@ -19,7 +20,8 @@ class WebDriverCoordsProviderTest {
         int y = 2;
         int height = 3;
         int width = 4;
-        when(element.getRect()).thenReturn(new Rectangle(x, y, height, width));
+        when(element.getLocation()).thenReturn(new Point(x, y));
+        when(element.getSize()).thenReturn(new Dimension(4, 3));
         Coords coords = webDriverCoordsProvider.ofElement(null, element);
         assertAll(
                 () -> assertEquals(x, coords.x),

